File: backend/src/application/csv/ProductListingCsvParser.ts
import { ProductListingInput } from "../../domain/entities/ProductListing.js";

export type CsvImportRowError = {
  row: number;
  errors: string[];
};

export type CsvParseResult = {
  rows: ProductListingInput[];
  errors: CsvImportRowError[];
  format: "shopify" | "simple";
};

type CsvRecord = Record<string, string>;

export class ProductListingCsvParser {
  parse(csv: string): CsvParseResult {
    const table = this.readCsv(csv);

    if (table.length === 0) {
      return { rows: [], errors: [{ row: 1, errors: ["CSV is empty."] }], format: "simple" };
    }

    const headers = table[0].map((header) => header.trim());
    const format = this.isShopify(headers) ? "shopify" : "simple";
    const rows: ProductListingInput[] = [];
    const errors: CsvImportRowError[] = [];

    table.slice(1).forEach((values, index) => {
      const rowNumber = index + 2;

      if (values.every((value) => !value.trim())) {
        return;
      }

      const record = this.toRecord(headers, values);
      const mapped = format === "shopify" ? this.mapShopify(record) : this.mapSimple(record);
      const rowErrors = this.validate(mapped, rowNumber);

      if (rowErrors.length > 0) {
        errors.push({ row: rowNumber, errors: rowErrors });
      } else {
        rows.push(mapped);
      }
    });

    return { rows, errors, format };
  }

  private readCsv(csv: string): string[][] {
    const rows: string[][] = [];
    let row: string[] = [];
    let cell = "";
    let quoted = false;

    for (let index = 0; index < csv.length; index += 1) {
      const char = csv[index];
      const next = csv[index + 1];

      if (char === "\"" && quoted && next === "\"") {
        cell += "\"";
        index += 1;
        continue;
      }

      if (char === "\"") {
        quoted = !quoted;
        continue;
      }

      if (char === "," && !quoted) {
        row.push(cell);
        cell = "";
        continue;
      }

      if ((char === "\n" || char === "\r") && !quoted) {
        if (char === "\r" && next === "\n") {
          index += 1;
        }
        row.push(cell);
        rows.push(row);
        row = [];
        cell = "";
        continue;
      }

      cell += char;
    }

    if (cell.length > 0 || row.length > 0) {
      row.push(cell);
      rows.push(row);
    }

    return rows.filter((cells) => cells.some((cellValue) => cellValue.trim().length > 0));
  }

  private isShopify(headers: string[]): boolean {
    const normalized = new Set(headers.map((header) => this.key(header)));
    return normalized.has("bodyhtml") || normalized.has("variantprice") || normalized.has("productcategory");
  }

  private toRecord(headers: string[], values: string[]): CsvRecord {
    return headers.reduce<CsvRecord>((record, header, index) => {
      record[this.key(header)] = values[index]?.trim() ?? "";
      return record;
    }, {});
  }

  private mapShopify(record: CsvRecord): ProductListingInput {
    const tags = this.splitTags(record.tags);
    const price = this.optionalNumber(record.variantprice || record.price);
    const inventoryQuantity = this.optionalInteger(record.variantinventoryqty || record.inventoryquantity);

    return {
      shopifyProductId: this.optional(record.id || record.productid || record.shopifyproductid),
      title: record.title?.trim() ?? "",
      handle: this.optional(record.handle),
      vendor: this.optional(record.vendor),
      productType: this.optional(record.type || record.producttype || record.productcategory),
      description: this.optional(record.bodyhtml || record.description),
      tags,
      price,
      currency: this.optional(record.currency) ?? "USD",
      inventoryQuantity,
      rawData: {
        importFormat: "shopify",
        sourceRow: record
      }
    };
  }

  private mapSimple(record: CsvRecord): ProductListingInput {
    return {
      shopifyProductId: this.optional(record.shopifyproductid || record.productid || record.id),
      title: record.title?.trim() ?? "",
      handle: this.optional(record.handle),
      vendor: this.optional(record.vendor),
      productType: this.optional(record.producttype),
      description: this.optional(record.description),
      tags: this.splitTags(record.tags),
      price: this.optionalNumber(record.price),
      currency: this.optional(record.currency) ?? "USD",
      inventoryQuantity: this.optionalInteger(record.inventoryquantity || record.inventory),
      rawData: {
        importFormat: "simple",
        sourceRow: record
      }
    };
  }

  private validate(input: ProductListingInput, rowNumber: number): string[] {
    const errors: string[] = [];

    if (!input.title.trim()) {
      errors.push("title is required");
    }

    if (input.price !== null && input.price !== undefined && Number.isNaN(input.price)) {
      errors.push("price must be a valid number");
    }

    if (
      input.inventoryQuantity !== null &&
      input.inventoryQuantity !== undefined &&
      !Number.isInteger(input.inventoryQuantity)
    ) {
      errors.push("inventoryQuantity must be an integer");
    }

    if (!/^[A-Za-z]{3}$/.test(input.currency ?? "USD")) {
      errors.push("currency must be a 3-letter code");
    }

    return errors.map((error) => `Row ${rowNumber}: ${error}`);
  }

  private splitTags(value: string | undefined): string[] {
    if (!value) {
      return [];
    }

    return [...new Set(value.split(/[|,]/).map((tag) => tag.trim()).filter(Boolean))];
  }

  private optional(value: string | undefined): string | null {
    const trimmed = value?.trim();
    return trimmed ? trimmed : null;
  }

  private optionalNumber(value: string | undefined): number | null {
    if (!value?.trim()) {
      return null;
    }

    const parsed = Number(value.replace(/[$,]/g, ""));
    return Number.isFinite(parsed) ? parsed : Number.NaN;
  }

  private optionalInteger(value: string | undefined): number | null {
    if (!value?.trim()) {
      return null;
    }

    const parsed = Number(value);
    return Number.isInteger(parsed) ? parsed : Number.NaN;
  }

  private key(value: string): string {
    return value.toLowerCase().replace(/[^a-z0-9]/g, "");
  }
}

File: app/services/product_listing_csv_parser.rb
class ProductListingCsvParser
  ParseResult = Struct.new(:rows, :errors, :format, keyword_init: true)
  RowError    = Struct.new(:row, :errors, keyword_init: true)

  def parse(csv)
    table = read_csv(csv)

    if table.empty?
      return ParseResult.new(rows: [], errors: [RowError.new(row: 1, errors: ["CSV is empty."])], format: "simple")
    end

    headers = table[0].map { |h| h.to_s.strip }
    format  = shopify_format?(headers) ? "shopify" : "simple"
    rows    = []
    errors  = []

    table[1..].each_with_index do |values, idx|
      row_number = idx + 2
      next if values.all? { |v| v.to_s.strip.empty? }

      record = headers.zip(values).to_h { |k, v| [normalize_key(k), v.to_s.strip] }
      mapped = format == "shopify" ? map_shopify(record) : map_simple(record)
      row_errors = validate(mapped, row_number)

      if row_errors.any?
        errors << RowError.new(row: row_number, errors: row_errors)
      else
        rows << mapped
      end
    end

    ParseResult.new(rows: rows, errors: errors, format: format)
  end

  private

  def read_csv(csv)
    rows    = []
    row     = []
    cell    = ""
    quoted  = false
    chars   = csv.chars

    chars.each_with_index do |char, i|
      nxt = chars[i + 1]

      if char == '"' && quoted && nxt == '"'
        cell += '"'
        next
      elsif char == '"'
        quoted = !quoted
      elsif char == ',' && !quoted
        row << cell; cell = ""
      elsif (char == "\n" || char == "\r") && !quoted
        row << cell; rows << row; row = []; cell = ""
      else
        cell += char
      end
    end

    row << cell
    rows << row if row.any?

    rows.reject { |r| r.all? { |c| c.strip.empty? } }
  end

  def shopify_format?(headers)
    normalized = headers.map { |h| normalize_key(h) }.to_set
    normalized.include?("bodyhtml") || normalized.include?("variantprice") || normalized.include?("productcategory")
  end

  def map_shopify(record)
    {
      shopify_product_id: optional(record["id"] || record["productid"] || record["shopifyproductid"]),
      title:              record["title"].to_s.strip,
      handle:             optional(record["handle"]),
      vendor:             optional(record["vendor"]),
      product_type:       optional(record["type"] || record["producttype"] || record["productcategory"]),
      description:        optional(record["bodyhtml"] || record["description"]),
      tags:               split_tags(record["tags"]),
      price:              optional_decimal(record["variantprice"] || record["price"]),
      currency:           optional(record["currency"]) || "USD",
      inventory_quantity: optional_integer(record["variantinventoryqty"] || record["inventoryquantity"]),
      raw_data:           { import_format: "shopify" }
    }
  end

  def map_simple(record)
    {
      shopify_product_id: optional(record["shopifyproductid"] || record["productid"] || record["id"]),
      title:              record["title"].to_s.strip,
      handle:             optional(record["handle"]),
      vendor:             optional(record["vendor"]),
      product_type:       optional(record["producttype"]),
      description:        optional(record["description"]),
      tags:               split_tags(record["tags"]),
      price:              optional_decimal(record["price"]),
      currency:           optional(record["currency"]) || "USD",
      inventory_quantity: optional_integer(record["inventoryquantity"] || record["inventory"]),
      raw_data:           { import_format: "simple" }
    }
  end

  def validate(input, row_number)
    errors = []
    errors << "Row #{row_number}: title is required"                       if input[:title].blank?
    errors << "Row #{row_number}: price must be a valid number"            if input[:price] == :invalid
    errors << "Row #{row_number}: inventoryQuantity must be an integer"    if input[:inventory_quantity] == :invalid
    errors << "Row #{row_number}: currency must be a 3-letter code"        unless (input[:currency] || "USD").match?(/\A[A-Za-z]{3}\z/)
    errors
  end

  def split_tags(value)
    return [] if value.blank?
    value.split(/[|,]/).map(&:strip).reject(&:blank?).uniq
  end

  def optional(value)
    value.to_s.strip.presence
  end

  def optional_decimal(value)
    return nil if value.blank?
    parsed = value.to_s.gsub(/[$,]/, "").to_f
    parsed.finite? ? parsed : :invalid
  end

  def optional_integer(value)
    return nil if value.blank?
    parsed = value.to_s.to_i
    value.to_s.strip.match?(/\A-?\d+\z/) ? parsed : :invalid
  end

  def normalize_key(value)
    value.to_s.downcase.gsub(/[^a-z0-9]/, "")
  end
end
